### ISSUE TITLE
zsysd: deadlock in mutating operations

### ISSUE THREAD

**Issue by aglasgall** [ubuntu/zsys#192]
**Describe the bug**
zsysd deadlock causes zsysctl operations to time out and fail

**To Reproduce**
This seems to happen whenever I try to do zsys operations that mutate state (e.g. zsysctl state remove)

**Expected behavior**
zsysd should not deadlock

**For ubuntu users, please run and  copy the following:**
1. `ubuntu-bug zsys --save=/tmp/report`
2. Copy paste below `/tmp/report` content:
```
Attached; too large to paste inline
```

**Installed versions:**
 - OS: Ubuntu 20.10 ("Groovy Gorilla")
- Zsysd running version: 0.5.2

**Additional context**

[report.txt](https://github.com/ubuntu/zsys/files/5992276/report.txt)

I'm not super familiar with Go, but I've attached a debugger to zsysd and poked around some. As far as I can tell, what seems to be happening most of the time is that a goroutine tries to take the RWRequest lock (e.g. https://github.com/ubuntu/zsys/blob/master/internal/daemon/state.go#L101) and hangs indefinitely because another goroutine is already holding the lock. It's not clear to me how to go about tracking down who's currently holding it.

I've tried turning verbosity up to -vv (debug), as you can see, but it hasn't been much help beyond what debugging already told me.

Clearly zsys state mutation worked at some point, as I now have 60+ states on this machine which was installed fresh last week; I'm not sure what's going on now.

**Comment by aglasgall** [comment]
I've dug into this a little more and it might not be a deadlock _per se_ - it looks like a goroutine that's trying to run a state GC is stuck waiting on a channel and that's blocking everything else:

```
(dlv) bt
 0  0x000000000043ebf0 in runtime.gopark
    at runtime/proc.go:305
 1  0x000000000040cd9d in runtime.chansend
    at runtime/chan.go:236
 2  0x000000000040cb65 in runtime.chansend1
    at runtime/chan.go:127
 3  0x0000000000920c25 in github.com/ubuntu/zsys/vendor/github.com/bicomsystems/go-libzfs.(*Dataset).Clones
    at github.com/ubuntu/zsys/vendor/github.com/bicomsystems/go-libzfs/zfs.go:853
 4  0x000000000092db54 in github.com/ubuntu/zsys/internal/zfs/libzfs.dZFSAdapter.Clones
    at <autogenerated>:1
 5  0x000000000093298f in github.com/ubuntu/zsys/internal/zfs.(*Dataset).checkNoClone
    at github.com/ubuntu/zsys/internal/zfs/helpers.go:322
 6  0x000000000093c253 in github.com/ubuntu/zsys/internal/zfs.(*NoTransaction).Destroy
    at github.com/ubuntu/zsys/internal/zfs/zfs.go:722
 7  0x000000000096369d in github.com/ubuntu/zsys/internal/machines.(*State).remove
    at github.com/ubuntu/zsys/internal/machines/state.go:369
 8  0x000000000094a1fe in github.com/ubuntu/zsys/internal/machines.(*Machines).GC
    at github.com/ubuntu/zsys/internal/machines/gc.go:227
 9  0x000000000097ce15 in github.com/ubuntu/zsys/internal/daemon.(*Server).GC
    at github.com/ubuntu/zsys/internal/daemon/service.go:183
10  0x00000000008c27a6 in github.com/ubuntu/zsys.(*ZsysLogServer).GC
    at github.com/ubuntu/zsys/zsys.streamlogger.go:656
11  0x00000000008bf369 in github.com/ubuntu/zsys._Zsys_GC_Handler
    at github.com/ubuntu/zsys/zsys.pb.go:2651
12  0x00000000008b0d59 in github.com/ubuntu/zsys/internal/streamlogger.ServerIdleTimeoutInterceptor
    at github.com/ubuntu/zsys/internal/streamlogger/server.go:73
13  0x0000000000898ae5 in github.com/ubuntu/zsys/vendor/google.golang.org/grpc.(*Server).processStreamingRPC
    at github.com/ubuntu/zsys/vendor/google.golang.org/grpc/server.go:1244
14  0x000000000089aae6 in github.com/ubuntu/zsys/vendor/google.golang.org/grpc.(*Server).handleStream
    at github.com/ubuntu/zsys/vendor/google.golang.org/grpc/server.go:1317
15  0x00000000008a8441 in github.com/ubuntu/zsys/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1
    at github.com/ubuntu/zsys/vendor/google.golang.org/grpc/server.go:722
16  0x000000000046eb81 in runtime.goexit
    at runtime/asm_amd64.s:1373
```

Not quite sure what exactly it's waiting on and why whatever it's waiting on isn't happening - like I said, I'm not super familiar with Go yet. I'll keep investigating as best I can, but some guidance would definitely be appreciated.

**Comment by aglasgall** [comment]
Upon further investigation, the checkNoClone frame is somewhat suspicious:

```
   321:		// TODO: this reopens the pool entirely, so can be a little bit slow. Could be reimplemented ourselves.
=> 322:		clones, err := d.dZFS.Clones()
```

I'm not sure why it's being more than "a little bit slow" here.

**Comment by aglasgall** [comment]
This is an incredibly dumb hack, but I patched go-libzfs to double the size of the channel buffer in Clones() and that seems to be getting me unstuck for now. I'm not sure why I had quite so many snapshots (over 2000 datasets total!) - I haven't manually taken any or done anything particularly odd beyond just installing packages while setting up my new laptop.

**Comment by aglasgall** [comment]
Okay, I'm putting this down for tonight since my immediate problem is solved, but it seems to me that "hangs completely if you have too many datasets" is rather undesirable behavior - being slow would be one thing, but this wasn't just slow, it was stuck waiting on a channel that nothing was apparently draining from. I've investigated the first half of that, but not the second.

**Comment by didrocks** [comment]
Thanks for the detailed analyzes!

We saw that go-libzfs was slow with too many datasets and this is why we had a plan to rewrite it, thinking the issue was in the C <-> Go back and forth. I’m surprised there can be a deadlock, but if the channel buffer is full, indeed, as those are treated in a single Goroutine, the deadlock may happen.

As you are in a state with a lot of datasets (depending on the number of zfs datasets and our hierachy to separate system and user datasets, with fine granularity for persistent ones, that can adds up quickly), do you have any feedback on the speed of computing them all for you?

I’m happy to do such a hack to unblock current situation for people in a similar case than you as a mitigation for now (as the rewrite of the C Go-libzfs handling may take a while).

Thanks again for the deep look into this and sorry you got into that state

**Comment by aglasgall** [comment]
Once I doubled the buffer size, zsysd chewed through the whole GC in a few minutes and everything seems to be working normally now, which feels entirely reasonable. I'm not sure how the channel buffer ended up full with nothing draining it (before patching, I left zsysd running with a debugger attached and sampled the buffer depth periodically; it never changed), though.

**Comment by didrocks** [comment]
Thanks for confirming and reaching out! We’ll apply the changes and backport it to the LTS.

**Comment by aglasgall** [comment]
For convenience, here's the patch:

```
diff -ur zsys-0.5.2/vendor/github.com/bicomsystems/go-libzfs/zfs.go zsys-0.5.2glasgall1/vendor/github.com/bicomsystems/go-libzfs/zfs.go
--- zsys-0.5.2/vendor/github.com/bicomsystems/go-libzfs/zfs.go	2020-09-28 08:39:07.000000000 -0400
+++ zsys-0.5.2glasgall1/vendor/github.com/bicomsystems/go-libzfs/zfs.go	2021-02-16 23:06:28.399589860 -0500
@@ -827,7 +827,7 @@
 	defer root.Close()
 	dIsSnapshot := d.IsSnapshot()
 	// USe breadth first search to find all clones
-	queue := make(chan Dataset, 1024)
+	queue := make(chan Dataset, 2048)
 	defer close(queue) // This will close and cleanup all
 	queue <- root      // start from the root element
 	for {
```

But I would be interested in knowing why the channel buffer was getting full and why whatever was supposed to be draining it wasn't.

**Comment by aglasgall** [comment]
OH. I see now and I understand your previous comment - this is just a straight up misuse of a channel as a normal queue, not being used as IPC at all.

**Comment by aglasgall** [comment]
If that's the case, this is a fundamental flaw in go-libzfs and doubling the channel buffer depth from 1024 to 2048 is only buying time; someone will probably still hit the limit - once you hit 1024 (or 2048, or 4096, or however many) datasets, any operations that call Clones() will hang indefinitely once the channel buffer fills.

**Comment by aglasgall** [comment]
Can we just replace the channel with a normal list or something that grows when full instead? That'd solve the problem without setting a hard upper limit on the number of datasets zsysd can handle without hanging mysteriously.

**Comment by didrocks** [comment]
Indeed, this is a flaw in go-libzfs. This is why this only buys us some time as you said.

FTR, we are not upstream for this binding (we only forked it and upstreamed as much patches as possible), but my guess is that the consumer and writer are on the same goroutine. It needs investigation and a simple list (or just ensuring reader and writer are not done in the same place) would help us, I want to come with a simple test case for upstream first to validate the theory.

**Comment by aglasgall** [comment]
As far as I can tell, consumer and writer are on the same goroutine: the function in question is doing a breadth-first search of ZFS datasets to discover them all and using a channel as the queue for the BFS algorithm.

**Comment by didrocks** [comment]
That makes total sense. It seems you are almost motivated to provide a patch upstream (FYI: https://github.com/bicomsystems/go-libzfs). They are reviewing and merging regularly our proposal. :)

**Comment by aglasgall** [comment]
Hah, fair enough :) If I have the time I'll look into it. In the meantime, doubling (or quadrupling) the channel buffer size should at least make it less likely that users will run into trouble in the short term. We might want to take a closer look at go-libzfs and see if this pattern occurs elsewhere, but I recognize that everyone has finite time available and that this will at least fix the immediate problem in the short term.

I'd open a PR with my one-line patch, but zsysd seems to be 3.0 (native), so there's no easy way to drop stuff in debian/patches and I don't know how patching vendored go packages is supposed to work, so I think I'll let you take it from here if that's OK

**Comment by didrocks** [comment]
Sure!
I’ll cherry-pick your one-line patch and merge it in our fork of go-libzfs as the dependency is vendorized during package source creation: https://github.com/ubuntu/zsys/blob/master/go.mod#L30 (or feel free to propose it there, as you wish)

Thanks again for the excellent debugging/digging work :)

**Comment by aglasgall** [comment]
I'm just glad I was able to figure out how to get my system unstuck! Thanks for being so responsive.

I couldn't figure out what branch in go-libzfs this should go on (there doesn't even seem to be a tag for 0.2.2, which is what zsys seems to be building against), and I'm well past the amount of time I set aside for this from my actual job, so I'll let you do the cherrypick. Good luck with the eventual rewrite!

**Comment by didrocks** [comment]
Thanks again, and no worry, will do!

**Comment by aglasgall** [comment]
Hi, just a gentle nudge on this - I do think that at least the "increase the channel buffer depth" workaround should make it into Hirsute, and it's already early April.

**Comment by didrocks** [comment]
Thanks for the nudge! Yeah, I wanted to wait a little bit if there was anything else we can do on time for hirsute, but it seems we won’t.

Adding your patch then and releasing to hirsute right now!

**Comment by didrocks** [comment]
released in v0.5.8
